### PR TITLE
Refactor OCP NLP interface

### PR DESCRIPTION
**Pull request by nielsvd**
Completely refactored NLP interface, summary of changes and TODOs:

## Changes
*  Separate computation of sensitivity information from SQP algorithm.
   * Introduce framework for so-called *sensitivity method*s. Different sensitivity methods could include: GGN, Exact-Hessian, IN/INIS, Multi-level methods, DOPUS, etc. These methods have a common interface to SQP, but can have their own memory, workspace, arguments.
   * An SQP iteration is reduced to: 
      1. Call sensitivity method (external routine).
      2. Call QP solver (external routine).
      3. Update optimization variables, which could at a later stage include an (external) globalization method.
   * The memory/workspace allocation routines are rewritten to mirror workflow of ``ocp_qp``-interface.
* New casadi_wrapper interface:
   * Assumes three inputs (states x, inputs u, parameters p) and three outputs (y, jacobian of y, hessian of y). Two side-notes:
      * In case you generate casadi c-code with only 1 or 2 inputs (instead of 3), it implies the function you wrote depends on only (x) or (x,u), respectively.
      * You can disable the computation of jacobians and hessians (regardless of whether code is generated for these outputs) by setting flags in the casadi_wrapper_in struct. This to save computation time (e.g., don't compute sensitivities if not necessary) and/or to reduce code size (e.g., don't generate second order sensitivities for a Hessian-free algorithm).
   * Compatible with sparse outputs.
   * Allocates workspace for casadi c-code.
   * The ``casadi_wrapper`` that was written for the simulator is renamed ``sim_casadi_wrapper``.
* The chain example is rewritten to comply with new workflow.

## TODOs (before merging)
[x] Resolve hack ``ocp_nlp_sqp.c:369``
[x] Re-implement SWIG interface.
[x] Re-implement unit tests.
[x] Implement IN/INIS with old interface.
[x] Implement freezeSens with (close to) old interface.

## Future TODOs
[ ] Re-implement IN/INIS (``ocp_nlp_sm_in``)?
[ ] Re-implement Frozen IN/INIS.
[ ] Write additional examples, with ones that have nonlinear bounds.

### CODE REVIEW

**Comment by zanellia** [comment]
We might have to wait until this https://github.com/acados/acados/pull/139 is merged. Quite some changes have been implemented in the ocp_nlp module to implement a GGN. Only thing missing is a fix to the hpmpc interface.

**Comment by roversch** [comment]
@zanellia what do you mean by 'fix'? Something minor?

**Comment by zanellia** [comment]
@roversch have a look at the discussion in https://github.com/acados/acados/pull/139

**Comment by nielsvd** [comment]
@zanellia, can you summarize the changes to the ``ocp_nlp`` that you made? I see #139 contains many changes not directly affecting the NLP interface.

File: acados/ocp_nlp/allocate_ocp_nlp.c [hunk 1]
```diff
@@ -73,165 +73,99 @@ static void free_ocp_nlp_in_bounds(ocp_nlp_in *const nlp) {
...
-    d_zeros(&out->lam[N], 2 * (in->nb[N] + in->nc[N] + in->ng[N]), 1);
+    d_zeros(&out->lam[N], 2 * (in->nb[N] + in->ng[N]), 1);
 }
 
 void free_ocp_nlp_out(int_t N, ocp_nlp_out *out) {
     for (int_t k = 0; k < N; k++) {
         d_free(out->x[k]);
         d_free(out->u[k]);
-        d_free(out->pi[k]);
+        d_free(out->pi[k + 1]);
```
[the remark this answers is not in the document]
Comment (nielsvd) [reply acados/ocp_nlp/allocate_ocp_nlp.c]: In this loop it's bug. Pi should be freed for indices 0...N-1.

File: acados/ocp_nlp/allocate_ocp_nlp.c [hunk 2]
```diff
@@ -148,12 +148,12 @@ void allocate_ocp_nlp_out(ocp_nlp_in *const in, ocp_nlp_out *out) {
     int_t N = in->N;
     out->x = (real_t **)calloc(N + 1, sizeof(*out->x));
     out->u = (real_t **)calloc(N + 1, sizeof(*out->u));
-    out->pi = (real_t **)calloc(N, sizeof(*out->pi));
+    out->pi = (real_t **)calloc(N + 1, sizeof(*out->pi));
```
[the remark this answers is not in the document]
Comment (nielsvd) [reply acados/ocp_nlp/allocate_ocp_nlp.c]: The last element in the array is actually of length zero (could be a NULL pointer as well actually). The only reason is to keep all arrays of the same length. I assume that the length of ``u`` is also N+1 for that reason, but correct me if I'm wrong.

I just pushed a commit which should introduce consistency in assigning values to ``pi[N]``, assigning a zero-length array and a NULL pointer were mixed.

File: acados/ocp_nlp/allocate_ocp_nlp.h [hunk 3]
```diff
@@ -27,8 +27,8 @@ extern "C" {
 #include "acados/ocp_nlp/ocp_nlp_common.h"
 #include "acados/utils/types.h"
 
-void allocate_ocp_nlp_in(int_t N, int_t *nx, int_t *nu, int_t *nb, int_t *nc,
-                         int_t *ng, ocp_nlp_in *const nlp);
+void allocate_ocp_nlp_in(int_t N, int_t *nx, int_t *nu, int_t *nb, int_t *ng,
```
[the remark this answers is not in the document]
Comment (nielsvd) [reply acados/ocp_nlp/allocate_ocp_nlp.h]: I removed it because it wasn't properly implemented in the previous interface. Since nonlinear constraints can cover affine constraints, usability is not affected. I believe that omitting affine constraints, at least for now, relieves some of the burden implementing different sensitivity methods.

File: acados/ocp_nlp/ocp_nlp_common.c [hunk 4]
```diff
@@ -19,73 +19,164 @@
 
 #include "acados/ocp_nlp/ocp_nlp_common.h"
 
+// TODO(nielsvd): only perform assert in debug mode?
```
[the remark this answers is not in the document]
Comment (nielsvd) [reply acados/ocp_nlp/ocp_nlp_common.c]: I wasn't aware of that, I removed the TODOs.

File: acados/ocp_nlp/ocp_nlp_sm_gn.c [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+    ocp_nlp_ls_cost *ls_cost = (ocp_nlp_ls_cost *)sm_in->cost;
+    ocp_nlp_function **path_constraints = sm_in->path_constraints;
+
+    for (int_t i = 0; i < N; i++) {
+        // Adjoint-based gradient correction (used for)
+        // TODO(nielsvd): create new sensitivity methods for inexact newton methods
+        sim_RK_opts *sim_opts = (sim_RK_opts *)sim[i]->args;
+        if (mem->inexact_init) {
+            if (sim_opts->scheme.type != exact) {
+                sim[i]->in->sens_adj = 1;
```
[the remark this answers is not in the document]
Comment (nielsvd) [reply acados/ocp_nlp/ocp_nlp_sm_gn.c]: OK, bools are not a native type to C. If we choose to support it, I suggest we add ``#include <stdbool.h>`` to ``acados/utils/types.h``. A commit below uses bool, and includes ``<stdbool.h>`` only there where explicitly needed.

File: acados/utils/casadi_wrapper.h [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+#endif
+
+#include "acados/utils/types.h"
+
+typedef struct {
+    const real_t *x;
+    const real_t *u;
+    const real_t *p;
+
+    int_t compute_jac;
```
[the remark this answers is not in the document]
Comment (nielsvd) [reply acados/utils/casadi_wrapper.h]: See comment above.

File: examples/c/CMakeLists.txt [hunk 7]
```diff
@@ -76,18 +80,18 @@ if(NOT CMAKE_CXX_COMPILER_ID MATCHES "MSVC")
 #     COMMAND "${CMAKE_COMMAND}" -E chdir ${EXECUTABLE_OUTPUT_PATH} ./mass_spring_partial_tightening_example
 # )
 
-    add_executable(chen_nmpc_qpoases_example
-        chen_nmpc_qpoases.c ${CHEN_MODEL_SRC})
-    target_link_libraries(chen_nmpc_qpoases_example acados)
-    add_test(chen_nmpc_qpoases_example chen_nmpc_qpoases_example)
+    # add_executable(chen_nmpc_qpoases_example
```
[the remark this answers is not in the document]
Comment (nielsvd) [reply examples/c/CMakeLists.txt]: For no good reason, I re-enabled them in a commit below.
